File: Daily-Streak/Day-15_Maximize-Area-of-Square-Hole-in-Grid/Solution.cpp
class Solution {
public:
    int maximizeSquareHoleArea(int n, int m, vector<int>& hBars, vector<int>& vBars) {

        // Sort removable bars
        sort(hBars.begin(), hBars.end());
        sort(vBars.begin(), vBars.end());

        // Find longest consecutive run in horizontal bars
        int maxH = 1, curr = 1;
        for (int i = 1; i < hBars.size(); i++) {
            if (hBars[i] == hBars[i - 1] + 1) {
                curr++;
            } else {
                curr = 1;
            }
            maxH = max(maxH, curr);
        }

        // Find longest consecutive run in vertical bars
        int maxV = 1;
        curr = 1;
        for (int i = 1; i < vBars.size(); i++) {
            if (vBars[i] == vBars[i - 1] + 1) {
                curr++;
            } else {
                curr = 1;
            }
            maxV = max(maxV, curr);
        }

        // Side length of largest square hole
        int side = min(maxH, maxV) + 1;

        return side * side;
    }
};

/*
🧪 Standalone intuition:
- Removing k consecutive bars creates a gap of size k+1
- Square side limited by smaller of horizontal & vertical gaps
*/

File: Daily-Streak/Day-15_Maximize-Area-of-Square-Hole-in-Grid/Solution.py
class Solution:
    def maximizeSquareHoleArea(self, n: int, m: int, hBars: list[int], vBars: list[int]) -> int:

        hBars.sort()
        vBars.sort()

        # Longest consecutive horizontal bars
        maxH = curr = 1
        for i in range(1, len(hBars)):
            if hBars[i] == hBars[i - 1] + 1:
                curr += 1
            else:
                curr = 1
            maxH = max(maxH, curr)

        # Longest consecutive vertical bars
        maxV = curr = 1
        for i in range(1, len(vBars)):
            if vBars[i] == vBars[i - 1] + 1:
                curr += 1
            else:
                curr = 1
            maxV = max(maxV, curr)

        side = min(maxH, maxV) + 1
        return side * side

"""
🧠 Why this works:
- k consecutive removed bars → gap of k+1
- Square limited by smaller dimension
- Elegant greedy + sorting
"""
